nn/conformer.py

```diff
@@ -0,0 +1,217 @@
+    out = nn.merge_dims(x, axes='static')
+    return out
+
+
+class ConformerEncoderLayer(nn.Module):
+  """
+  Represents a conformer block
+  """
+
+  def __init__(
+        self, conv_kernel_size: int, activation_ff: Callable[[nn.LayerRef], nn.LayerRef], dim_ff: int, dropout: float,
+        att_dropout: float, enc_key_dim: int, num_heads: int, batch_norm_opts: Optional[Dict[str, Any]] = None):
+    """
+    :param conv_kernel_size:
+    :param activation_ff:
+    :param ff_dim:
+    :param dropout:
+    :param att_dropout:
+    :param enc_key_dim:
+    :param num_heads:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+
+    self.ffn1 = _PositionwiseFeedForward(
+      out_dim=enc_key_dim, dim_ff=dim_ff, dropout=dropout, activation=activation_ff)
+
+    self.ffn2 = _PositionwiseFeedForward(
+      out_dim=enc_key_dim, dim_ff=dim_ff, dropout=dropout, activation=activation_ff)
+
+    self.conv_block = _ConformerConvBlock(
+      out_dim=enc_key_dim, kernel_size=conv_kernel_size, batch_norm_opts=batch_norm_opts)
+
+    self.self_att = MultiheadAttention(enc_key_dim, num_heads, dropout=att_dropout)  # TODO: to be implemented
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    # FFN
+    x_ffn1_ln = nn.layer_norm(inp)
+    x_ffn1 = self.ffn1(x_ffn1_ln)
+    x_ffn1_out = 0.5 * nn.dropout(x_ffn1, dropout=self.dropout) + inp
+
+    # MHSA
+    x_mhsa_ln = nn.layer_norm(x_ffn1_out)
+    x_mhsa = self.self_att(x_mhsa_ln)
+    x_mhsa_out = x_mhsa + x_ffn1_out
+
+    # Conv
+    x_conv_ln = nn.layer_norm(x_mhsa_out)
+    x_conv = self.conv_block(x_conv_ln)
+    x_conv_out = nn.dropout(x_conv, dropout=self.dropout) + x_mhsa_out
+
+    # FFN
+    x_ffn2_ln = nn.layer_norm(x_conv_out)
+    x_ffn2 = self.ffn2(x_ffn2_ln)
+    x_ffn2_out = 0.5 * nn.dropout(x_ffn2, dropout=self.dropout) + x_conv_out
+
+    # last LN layer
+    return nn.layer_norm(x_ffn2_out)
+
+
+class ConformerEncoder(nn.Module):
+  """
+  Represents Conformer encoder architecture
+  """
+
+  def __init__(
+        self, encoder_layer: nn.Module, num_blocks: int, conv_kernel_size: int = 32,
+        activation_ff: Callable[[nn.LayerRef], nn.LayerRef] = nn.swish, dim_ff: int = 512, dropout: float = 0.1,
+        att_dropout: float = 0.1, enc_key_dim: int = 256, num_heads: int = 4,
+        batch_norm_opts: Optional[Dict[str, Any]] = None):
+    """
+    :param encoder_layer:
+    :param num_blocks:
+    :param conv_kernel_size:
+    :param ff_act:
+    :param ff_dim:
+    :param dropout:
+    :param att_dropout:
+    :param enc_key_dim:
+    :param att_n_heads:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+
+    self.conv_subsample_layer = _ConformerConvSubsample(
+      filter_sizes=[(3, 3), (3, 3)], pool_sizes=[(2, 2), (2, 2)], channel_sizes=[enc_key_dim, enc_key_dim],
+      dropout=dropout)
+
+    self.linear = nn.Linear(n_out=enc_key_dim, with_bias=False)
+
+    self.conformer_blocks = nn.Sequential(
```

Comment:
Sorry, maybe I suggested it badly earlier, I think we should call it `layers` here instead, to be consistent with `encoder_layer`. Also no need to explicitly put the `conformer_` prefix. The class already is named like that.
```suggestion
    self.layers = nn.Sequential(
```

Reply:
Ok.